taichi/backends/cuda/cuda_device.cpp

```diff
@@ -0,0 +1,43 @@
+#include "taichi/backends/cuda/cuda_device.h"
+
+TLANG_NAMESPACE_BEGIN
+
+namespace cuda {
+
+CudaDevice::AllocInfo CudaDevice::get_alloc_info(DeviceAllocation handle) {
+  validate_device_alloc(handle);
+  return allocations_[handle.alloc_id];
+}
+
+DeviceAllocation CudaDevice::allocate_memory(const AllocParams &params) {
```

Comment:
Note that there is another memory allocation mechanism in Taichi, where you pre-allocate a huge chunk of memory and use it as the pool. We should think about how to better unify these two approaches.

Reply:
I agree. 

Currently, the `CudaDevice::allocate_memory` is used by the memory pool when it requires memory from the cuda driver. This is the exact opposite as our vulkan device, whose `VulkanDeivce::allocate_memory` uses an underlying memory pool.

Reply:
How about we move the cuda memory pool inside the device api, so that Device APIs `allocate_memory` can always be used as-is?

Reply:
https://github.com/NVIDIA/cnmem Looks interesting. There isn't a VulkanMemoryAllocator equivalent in CUDA i think. (There is an DX12MemoryAllocator lol)

Reply:
> How about we move the cuda memory pool inside the device api in the future, so that Device APIs allocate_memory can always be used as-is?

SG, right now https://github.com/taichi-dev/taichi/blob/e5914a458776bb921dd95cb9a00126dafe675b7a/taichi/system/snode_tree_buffer_manager.cpp#L43-L48 kind of achieves what we want. (It invokes the allocation function that is JIT compiled from LLVM, which makes things more complicated than necessary.) FYI @qiao-bo

Reply:
> It invokes the allocation function that is JIT compiled from LLVM, which makes things more complicated than necessary

I should have clarified here: The LLVM module needs to have a function for memory allocation, because the sparse SNode system needs that. This part is completely fine. However, I feel like it's simpler to implement the entire memory allocation functionality on the host side, then passing in a function pointer into the JIT module, rather than the other way around. Right now the situation is a bit complicated:

* If CUDA unified memory is enabled, then we are passing in the function pointer to the JIT module: https://github.com/taichi-dev/taichi/blob/e5914a458776bb921dd95cb9a00126dafe675b7a/taichi/llvm/llvm_program.cpp#L345-L350
* Otherwise, the memory allocation is done completely inside the JIT module: https://github.com/taichi-dev/taichi/blob/e5914a458776bb921dd95cb9a00126dafe675b7a/taichi/runtime/llvm/runtime.cpp#L783-L814

Reply:
That makes sense, let's aim to only implement memory allocation on the host, and always pass a func ptr to LLVMRuntime.